react/components/GoogleMap.tsx

```diff
@@ -0,0 +1,373 @@
+ */
+const PAN_LEFT_LNG = -30
+
+const MAP_STYLES: CSSProperties = {
+  height: '100%',
+  width: '100%',
+  position: 'absolute',
+  top: 0,
+  zIndex: 0,
+}
+
+/**
+ * Transforms the geoCoordinates of the API to a LatLng object from Google Maps
+ * SDK
+ *
+ * @param {[number, number]} geoCoordinates - Geocoordinates in [longitude, latitude] format
+ * @returns {google.maps.LatLng} LatLng object with the same position as `geoCoordinates`
+ */
+const geoCoordinatesToLatLng = (geoCoordinates: [number, number]) => {
+  const [lng, lat] = geoCoordinates
+  const location = new google.maps.LatLng(lat, lng)
+
+  return location
+}
+
+const isGeoCoordinatesEqual = (
+  geoA: Array<number | undefined>,
+  geoB: number[]
+) => {
+  if (geoA.length !== geoB.length) {
+    return false
+  }
+
+  const [lngA, latA] = geoA
+  const [lngB, latB] = geoB
+
+  return lngA === lngB && latA === latB
+}
+
+const useMapEvent = (
+  map: google.maps.Map | null,
+  eventName: string,
+  handler: () => void
+) => {
+  const handlerRef = useRef(handler)
+
+  useEffect(() => {
+    handlerRef.current = handler
+  })
+
+  useEffect(() => {
+    if (map == null) {
+      return
+    }
+
+    const event = google.maps.event.addListener(map, eventName, () => {
+      handlerRef.current()
+    })
+
+    return () => {
+      google.maps.event.removeListener(event)
+    }
+  }, [eventName, map])
+}
+
+interface GoogleMapRefObject {
+  getBounds(): google.maps.LatLngBounds | undefined
+}
+
+export const GoogleMap = forwardRef<GoogleMapRefObject, Props>(
+  function GoogleMap(
+    {
+      children,
+      center = [],
+      isLargeScreen = false,
+      isLoading,
+      loadingElement,
+      onClick,
+      referenceCenter,
+      bounds,
+    },
+    ref
+  ) {
+    const {
+      setMapCenterLatLng,
+      setShouldSearchArea,
+      setActiveState,
+      setActiveSidebarState,
+      activeState,
+    } = useModalState()
+
+    const [map, setMap] = useState<google.maps.Map | null>(null)
+
+    const referenceCenterRef = useRef(referenceCenter)
+
+    useEffect(() => {
+      if (referenceCenter === referenceCenterRef.current) {
+        return
+      }
+
+      referenceCenterRef.current = referenceCenter
+    }, [referenceCenter])
+
+    const toggleSearchArea = useCallback(() => {
+      const currentCenter = map?.getCenter()
+
+      if (currentCenter == null || referenceCenterRef.current == null) {
+        return
+      }
+
+      const [referenceCenterLng, referenceCenterLat] =
+        referenceCenterRef.current
+
+      const differenceLat = (referenceCenterLat - currentCenter.lat()) * 1000
+      const differenceLng = (referenceCenterLng - currentCenter.lng()) * 1000
+
+      const DISTANCE = 50
+
+      const outerLat = differenceLat > DISTANCE || differenceLat < -DISTANCE
+      const outerLng = differenceLng > DISTANCE || differenceLng < -DISTANCE
+
+      setShouldSearchArea(outerLat || outerLng)
```

Comment:
Have you considered using `Math.abs`? I believe it makes the code easier to understand

Besides that, I would put the constant on the top like the other contestants and would also remove the `outer` constants and make the comparison inside the `setShouldSearchArea`, but that's more of a personal preference. Wdyt?

```suggestion
      const differenceLat = Math.abs(referenceCenterLat - currentCenter.lat()) * 1000
      const differenceLng = Math.abs(referenceCenterLng - currentCenter.lng()) * 1000

      setShouldSearchArea(differenceLat > DISTANCE || differenceLng > DISTANCE)
```

Reply:
makes sense to me! I haven't put much thought in this code because I didn't wanted to break something that was already working 😅

Reply:
ahh no worries. I thought that this particular file might have been created from scratch